### PR TITLE
KTOR-1094 Drop IoBuffer

### PR DESCRIPTION
**Pull request by e5l**

### CODE REVIEW

File: ktor-io/common/src/io/ktor/utils/io/core/BufferFactory.kt [hunk 1]
```diff
@@ -51,34 +51,33 @@ internal inline fun <R> withChunkBuffer(pool: ObjectPool<ChunkBuffer>, block: Ch
...
         instance.unlink()
     }
 
-    override fun validateInstance(instance: IoBuffer) {
+    override fun validateInstance(instance: ChunkBuffer) {
         super.validateInstance(instance)
 
-        if (instance === IoBuffer.Empty) {
-            error("IoBuffer.Empty couldn't be recycled")
+        if (instance === ChunkBuffer.Empty) {
```
Comment (whyoleg) [inline ktor-io/common/src/io/ktor/utils/io/core/BufferFactory.kt:76]: triple check for ChunkBuffer.Empty equality :)

File: ktor-io/common/src/io/ktor/utils/io/core/BufferPrimitives.kt [hunk 2]
```diff
@@ -24,9 +24,7 @@ public inline fun Buffer.forEach(block: (Byte) -> Unit) {
  */
 public fun Buffer.readUByte(): UByte = readByte().toUByte()
 
-@Suppress("EXTENSION_SHADOWED_BY_MEMBER", "DEPRECATION")
-@Deprecated("IoBuffer is deprecated. Use Memory or Input instead.")
-public inline fun IoBuffer.readUByte(): UByte = (this as Buffer).readUByte()
+public inline fun ChunkBuffer.readUByte(): UByte = (this as Buffer).readUByte()
```
Comment (whyoleg) [inline ktor-io/common/src/io/ktor/utils/io/core/BufferPrimitives.kt:27]: ChunkBuffer extends Buffer, are those functions needed for ChunkBuffer?

File: ktor-io/jvm/src/io/ktor/utils/io/core/ByteReadPacketExtensions.kt [hunk 3]
```diff
@@ -30,13 +30,11 @@ private class SingleByteBufferPool(
...
     override fun produceInstance(): ChunkBuffer {
-        @Suppress("DEPRECATION")
-        return IoBuffer(instance, this as ObjectPool<IoBuffer>)
+        return ChunkBuffer(instance, this)
     }
 
     override fun disposeInstance(instance: ChunkBuffer) {
-        @Suppress("DEPRECATION")
-        check(instance is IoBuffer) { "Only IoBuffer could be recycled" }
+        check(instance is ChunkBuffer) { "Only ChunkBuffer could be recycled" }
```
Comment (whyoleg) [inline ktor-io/jvm/src/io/ktor/utils/io/core/ByteReadPacketExtensions.kt:37]: redundant

File: ktor-io/common/src/io/ktor/utils/io/core/internal/ChunkBuffer.kt [hunk 4]
```diff
@@ -162,24 +154,21 @@ public open class ChunkBuffer internal constructor(
...
             }
         }
 
-        @Suppress("DEPRECATION")
+        public val Empty: ChunkBuffer = ChunkBuffer(Memory.Empty, null, EmptyPool)
+
         internal val NoPool: ObjectPool<ChunkBuffer> = object : NoPoolImpl<ChunkBuffer>() {
             override fun borrow(): ChunkBuffer {
-                return IoBuffer(DefaultAllocator.alloc(DEFAULT_BUFFER_SIZE), null, this as ObjectPool<IoBuffer>)
+                return ChunkBuffer(DefaultAllocator.alloc(DEFAULT_BUFFER_SIZE), null, this as ObjectPool<ChunkBuffer>)
```
Comment (whyoleg) [inline ktor-io/common/src/io/ktor/utils/io/core/internal/ChunkBuffer.kt:168]: cast is redundant

File: ktor-io/js/src/io/ktor/utils/io/core/ByteReadPacket.kt [hunk 5]
```diff
@@ -22,12 +22,12 @@ public actual fun ByteReadPacket(
     val pool = object : SingleInstancePool<ChunkBuffer>() {
         override fun produceInstance(): ChunkBuffer {
             @Suppress("DEPRECATION")
-            return IoBuffer(Memory.of(sub), null, this as ObjectPool<IoBuffer>)
+            return ChunkBuffer(Memory.of(sub), null, this as ObjectPool<ChunkBuffer>)
```
Comment (whyoleg) [inline ktor-io/js/src/io/ktor/utils/io/core/ByteReadPacket.kt:25]: cast is redundant

File: ktor-io/js/src/io/ktor/utils/io/core/ByteReadPacket.kt [hunk 6]
```diff
@@ -22,12 +22,12 @@ public actual fun ByteReadPacket(
...
         }
 
         override fun disposeInstance(instance: ChunkBuffer) {
             @Suppress("DEPRECATION")
-            check(instance is IoBuffer) { "Only IoBuffer could be recycled" }
+            check(instance is ChunkBuffer) { "Only ChunkBuffer could be recycled" }
```
Comment (whyoleg) [inline ktor-io/js/src/io/ktor/utils/io/core/ByteReadPacket.kt:30]: redundant check

File: ktor-io/posix/src/io/ktor/utils/io/core/ByteReadPacket.kt [hunk 7]
```diff
@@ -29,13 +29,13 @@ public actual fun ByteReadPacket(
             pinned = content
 
             @Suppress("DEPRECATION")
-            return IoBuffer(Memory.of(base, length), null, this as ObjectPool<IoBuffer>)
+            return ChunkBuffer(Memory.of(base, length), null, this as ObjectPool<ChunkBuffer>)
```
Comment (whyoleg) [inline ktor-io/posix/src/io/ktor/utils/io/core/ByteReadPacket.kt:32]: cast is redundant

File: ktor-io/posix/src/io/ktor/utils/io/core/ByteReadPacket.kt [hunk 8]
```diff
@@ -29,13 +29,13 @@ public actual fun ByteReadPacket(
...
         }
 
         override fun disposeInstance(instance: ChunkBuffer) {
             check(pinned != null) { "The array hasn't been pinned yet" }
             @Suppress("DEPRECATION")
-            check(instance is IoBuffer) { "Only IoBuffer could be recycled" }
+            check(instance is ChunkBuffer) { "Only ChunkBuffer could be recycled" }
```
Comment (whyoleg) [inline ktor-io/posix/src/io/ktor/utils/io/core/ByteReadPacket.kt:38]: redundant check

**Review by whyoleg — COMMENTED** [review]
found some minor remarks looking through PR :)

File: ktor-io/common/src/io/ktor/utils/io/core/BufferPrimitives.kt (hunk 2 above)
Comment (e5l) [reply ktor-io/common/src/io/ktor/utils/io/core/BufferPrimitives.kt:27]: We will drop members to work with primitive types and leave extensions only

File: ktor-io/common/src/io/ktor/utils/io/core/BufferFactory.kt (hunk 1 above)
Comment (e5l) [reply ktor-io/common/src/io/ktor/utils/io/core/BufferFactory.kt:76]: Could you tell me why it can be a problem?

Comment (e5l) [reply ktor-io/common/src/io/ktor/utils/io/core/BufferFactory.kt:76]: Thanks, will fix

**Review by cy6erGn0m — APPROVED** [review]
Finally. But i would prefer to extract codestyle changes and version upgrade to separate prs. Also, we drop buffer and introduce new extensions at once..

**Pull request merged**